utils.go

```diff
@@ -281,3 +283,83 @@ func isEmptyDetails(details *unversioned.StatusDetails) bool {
 	}
 	return false
 }
+
+// withExponentialBackoff retries the specified function fn exponentially.
+// If fn returns an already exists error, the operation is retried - any other error
+// aborts the execution.
+// It expects fn to return errors converted to trace type hierarchy with ConvertError
+func withExponentialBackoff(fn func() error) error {
+	const initialDelay = 1 * time.Second
+	backoff := wait.Backoff{
+		Duration: initialDelay,
+		Factor:   2.0,
+		Steps:    10,
+	}
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		err := fn()
+		if err != nil && !trace.IsAlreadyExists(err) {
+			// abort
+			return false, trace.Wrap(err)
+		}
+		if trace.IsAlreadyExists(err) {
+			// retry
+			return false, nil
+		}
+		return true, nil
+	})
+	return trace.Wrap(err)
+}
+
+func deletePodsList(podIface corev1.PodInterface, pods []v1.Pod, entry log.Entry) error {
+	for _, pod := range pods {
+		entry.Debugf("deleting pod %v", pod.Name)
+		err := ConvertError(podIface.Delete(pod.Name, nil))
+		if err != nil && !trace.IsNotFound(err) {
+			return ConvertError(err)
+		}
+	}
+	return nil
```

Comment:
Is it on purpose that this method doesn't "waitForPods" and the one below does? Also, do we need two similar methods? This one is used in a single place and you could convert slice to a map there and reuse the one below.

Reply:
I thought it was not worth the trouble - this code is in bad need for a rewrite that would provide more flexibility when adding new types (it is a lot of work currently), so I'd rather have a good stab at restructuring this for a greater good at a later point.

Reply:
:+1: